app/services/foreman_templates/template_importer.rb

```diff
@@ -54,54 +51,52 @@ def parse_files!
         result_lines << 'Parsing: ' + template.gsub(/#{@dir}#{@dirname}/, '') if @verbose
 
         metadata = parse_metadata(text)
-        metadata['associate'] = @associate
 
         # Get the name and filter
-        filename = template.split('/').last
-        title    = filename.split('.').first
-        name     = metadata['name'] || title
-        name     = auto_prefix(name)
+        name = metadata['name']
+        name = auto_prefix(name)
         if @filter
           matching = name.match(/#{@filter}/i)
           matching = !matching if @negate
-          next unless matching
+          unless matching
+            result_lines << "Skipping template 'name' since it's name is matching filter condition" if @verbose
+            next
+          end
         end
 
+        options = { :force => @force, :associate => @associate, :lock => @lock }
+        template_type = 'Unknown template type'
         begin
-          # Expects a return of { :diff, :status, :result, :errors }
-          data = if metadata['model'].present?
-                   metadata['model'].constantize.import!(name, text, metadata, @force, @lock)
-                 else
-                   # For backwards-compat before "model" metadata was added
-                   case metadata['kind']
-                   when 'ptable'
-                     Ptable.import!(name, text, metadata, @force, @lock)
-                   when 'job_template'
-                     # TODO: update REX templates to have `model` and delete this
-                     update_job_template(name, text)
-                   else
-                     ProvisioningTemplate.import!(name, text, metadata, @force, @lock)
-                   end
-                 end
-
-          if data[:diff].nil? && data[:old].present? && data[:new].present?
-            data[:diff] = calculate_diff(data[:old], data[:new])
+          if metadata.key?('model')
+            template_type = metadata['model'].constantize
+          else
+            result_lines << "Metadata for template '#{name}' do not specify 'model', import failed"
+            next
+          end
+
+          template = template_type.import!(name, text, options)
+
+          data = {}
+          if template.template_changed?
```

Comment:
Will always be `false`. `template_type.import!` on l.77 returns newly saved template, no diffs are ever shown.

Reply:
bummer, `import!` seems to be too high-level, we need to hook into the process of saving to get the diff before save is called, that would actually help solving the problem with locking too, I'll look into it